### DISCUSSION TITLE
5장. 접근 거부 시 로그인 페이지로 라우팅하기. (P290)

### DISCUSSION THREAD

**Discussion by mthekid** [fsoftwareengineer/todo-application#59] [질문과 답변] [answered]
오류나 에러가 나는 페이지
=> P290

오류/에러 로그
=> 프론트에 call 메서드에서 fetch() 부분. error 값이 undefined로 나옵니다.
백엔드에서 403 응답 코드를 보내주지 않고 있습니다.

작성 중인 코드

백엔드 보안 파트 + 프론트 서버 통합 파트

오류/에러를 해결하기 위해 현재까지 시도 해 본 방법들

1. 포스트맨으로 백엔드에 별도로 요청을 보냈을 때 403 응답이 오지 않는 것을 확인했습니다.
2. 프론트 서버 측에서 error 객체를 핸들링할 때 값들을 콘솔로 확인했는데 defined가 나왔습니다. 

<현재 상황>

안녕하세요. P290에서 리액트 프론트 서버와 백엔드 서버를 연결해서 로그인 통합하기를 진행하고 있습니다.

App.js의 componentDidMount에서 get메서드로 /todo를 부분에서 해메고 있습니다.

apiService 컴포넌트의 call 메서드를 사용하는 과정에서

백엔드로 /todo 요청을 보내고 응답을 받아서 처리해야 하는데

fetch(options.url, options)에서 error(예외)를 처리하는 과정에서 "Unexpected end of JSON input" 에러가 잡힙니다.

그래서 error.status === 403까지 코드로 이어지지 않고 있습니다. 

포스트맨으로 백엔드에 /todo 요청을 따로 보냈는데 응답 결과가 오질 않는 것을 확인 했습니다. [시도방법1]
[ 이상한 것은 개발자 도구에서는 403 응답 처리가 되었다고는 나옵니다. ] 

P270을 보면 위조된 토큰의 경우 상태값이 403인 응답이 와야 하는데 백엔드에서는 403 응답을 보내주지 않고 있습니다.

백엔드 측에서 요청을 처리하는 과정에서 토큰이 위조된 경우에는 SignatureException을 던지고 토큰을 아예 전송하지 않는 경우는 따로 응답을 보내주지 않더군요. [ JwtAuthenticationFilter 파트 ]

책에서는 백엔드에 요청을 보내면 403(Forbidden) 응답 코드가 오는? 상황으로 진행되는 것 같은데, 백엔드 코드 어디에서 응답코드를 반환하는 것인지 모르겠습니다. 

백엔드 측에서 응답을 따로 보내지 않고 있으니 

React 프론트 서버에서 error를 처리하는 부분에서 undefined라는 결과가 나오게 되는 것 같습니다. [ 추측입니다. ] 

혹시 저와 같은 상황에 있는 분이 계신지 궁금합니다.

<참고로 예제 진행에서 Secret-key 부분 말고는 책의 내용 그대로를 따라했습니다. >

private static final SecretKey SECRET_KEY = Keys.secretKeyFor(SignatureAlgorithm.HS512); 로 key값을 두고 해당 키를 보안키로 사용했습니다. [ 콘솔에서 예외가 뜨고 권장하는데로 바꾸었습니다. ]

<리액트 프론트 서버 에러코드>

![image](https://user-images.githubusercontent.com/31848154/150285485-8ecca119-07e7-48f4-85f6-8a6128bcafcc.png)

<포스트맨 응답 반응>

- todo 요청을 보내면 Filter is running이라는 콘솔반응만 나옵니다. 별도의 응답 반응이 없습니다.

![image](https://user-images.githubusercontent.com/31848154/150285561-fb12618d-9a8b-4cd0-9db5-b35964277cfa.png)

필요한 정보가 있으시면 바로 피드백 하겠습니다 .감사합니다.

=> Discussion을 찾아보다가 onthesly 님의 p270, p291 인증관련 질문입니다 #30
을 읽고 해결했습니다.

스프링 부트 버전이 2.6.1이었는데 2.5.7로 바꾸고나서 바로 리다이렉트가 잘 진행되었습니다.

엄청 해맸는데..다행이네요.

**Comment by fsoftwareengineer** [comment, marked as answer]
버전을 여쭤보려고 했는데 찾으셨다니 다행이네요!

**Reply by mthekid** [reply]
답변 감사합니다!. 그리고 끝내주는 책도 감사합니다. 진짜... 찾고 있었던 내용이 모드 연계되어서 너무 행복합니다.
